src/lib/operators/table_scan.cpp

```diff
@@ -0,0 +1,133 @@
+  _appendPositionList(position_list, accessor, values.size(), compare_value, chunk_id, false);
+}
+
+template <typename T>
+void TableScan::TableScanImpl<T>::_scanColumn(std::shared_ptr<PosList> position_list,
+                                              std::shared_ptr<DictionaryColumn<T>> dc, const ChunkID chunk_id) {
+  const auto& id_values = dc->attribute_vector();
+  ValueID dict_id = INVALID_VALUE_ID;
+  bool all_false = false;
+  bool all_true = false;
+  const auto compare_value = opossum::type_cast<T>(_search_value);
+
+  switch (_scan_type) {
+    case ScanType::OpEquals: {
+      auto lower_bound = dc->lower_bound(compare_value);
+      dict_id = lower_bound;
+      all_false = lower_bound == dc->upper_bound(compare_value);
+      break;
+    }
+    case ScanType::OpNotEquals: {
+      auto lower_bound = dc->lower_bound(compare_value);
+      dict_id = lower_bound;
+      all_true = lower_bound == dc->upper_bound(compare_value);
+      break;
+    }
+    case ScanType::OpGreaterThan: {
+      dict_id = dc->upper_bound(compare_value);
+      _scan_type = ScanType::OpGreaterThanEquals;
+      all_false = dict_id == INVALID_VALUE_ID;
+      all_true = dict_id == ValueID(0);
+      break;
+    }
+    case ScanType::OpLessThanEquals: {
+      dict_id = dc->upper_bound(compare_value);
+      _scan_type = ScanType::OpLessThan;
+      all_false = dict_id == ValueID(0);
+      all_true = dict_id == INVALID_VALUE_ID;
+      break;
+    }
+    case ScanType::OpGreaterThanEquals: {
+      dict_id = dc->lower_bound(compare_value);
+      all_true = dict_id == ValueID(0);
+      all_false = dict_id == INVALID_VALUE_ID;
+      break;
+    }
+    case ScanType::OpLessThan: {
+      dict_id = dc->lower_bound(compare_value);
+      all_true = dict_id == INVALID_VALUE_ID;
+      all_false = dict_id == ValueID(0);
+      break;
+    }
+  }
+
+  if (!all_false) {
+    const auto& accessor = [&](size_t chunk_offset) { return id_values->get(chunk_offset); };
+
+    _appendPositionList(position_list, accessor, dc->size(), dict_id, chunk_id, all_true);
+  }
+}
+
+template <typename T>
+void TableScan::TableScanImpl<T>::_scanColumn(std::shared_ptr<PosList> position_list,
+                                              std::shared_ptr<ReferenceColumn> rc,
+                                              std::shared_ptr<const Table>& referenced_table, const ChunkID chunk_id) {
+  const auto& column_pos_list = rc->pos_list();
+  referenced_table = rc->referenced_table();
+  const auto compare_value = opossum::type_cast<T>(_search_value);
+
+  for (const auto& row_id : *column_pos_list) {
+    const T value = _evaluateReferenceColumn(rc, row_id);
+    if (_evaluate_scan(value, compare_value)) {
+      position_list->push_back(row_id);
+    }
+  }
+}
+
+template <typename T>
+T TableScan::TableScanImpl<T>::_evaluateReferenceColumn(std::shared_ptr<ReferenceColumn> rc, const RowID& row_id) {
```

Comment:
`_evaluateReferenceColumn` is not overly specific. This method returns values "behind" the ReferenceColumn. Maybe something like `_getValueFromReferenceColumn` would be more precise.